daemon/response.go

```diff
@@ -281,55 +282,142 @@ func (f fileResponse) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	e := make(chan error)
+	for _, r := range rr.readers {
+		go rr.logReader(r, c, e)
+	}
 
-		// ignore the error...
-		t, _ := log.Time()
-		if err = enc.Encode(client.Log{
-			Timestamp: t,
-			Message:   log.Message(),
-			SID:       log.SID(),
-			PID:       log.PID(),
-		}); err != nil {
-			break
+	writeError := func(err error) {
+		fmt.Fprintf(writer, `\x1E{"error": %q}\n`, err)
+		logger.Noticef("cannot stream response; problem reading: %v", err)
+	}
+
+	writeLogs := func(logs []systemd.Log) error {
+		// sort by timestamp ascending
+		sort.Slice(logs, func(i, j int) bool {
+			ti, _ := logs[i].Time()
+			tj, _ := logs[j].Time()
+			return ti.Before(tj)
+		})
+
+		for _, l := range logs {
+			writer.WriteByte(0x1E) // RS -- see ascii(7), and RFC7464
+
+			// ignore the error...
+			t, _ := l.Time()
+			if err := enc.Encode(client.Log{
+				Timestamp: t,
+				Message:   l.Message(),
+				SID:       l.SID(),
+				PID:       l.PID(),
+			}); err != nil {
+				return err
+			}
 		}
 
 		if rr.follow {
-			if e := writer.Flush(); e != nil {
-				break
+			if err := writer.Flush(); err != nil {
+				return nil
 			}
 			if hasFlusher {
 				flusher.Flush()
 			}
 		}
+		return nil
 	}
-	if err != nil && err != io.EOF {
-		fmt.Fprintf(writer, `\x1E{"error": %q}\n`, err)
-		logger.Noticef("cannot stream response; problem reading: %v", err)
+
+	var logReadersDone int
+	for {
+		var logs []systemd.Log
+
+		// always block read on the first one to ensure we don't waste any
+		// time spinning on a channel that will never have any logs
+		select {
+		case log := <-c:
+			logs = append(logs, log)
+		case err := <-e:
+			if err != io.EOF {
+				writeError(err)
+			}
+			logReadersDone++
+		}
+
+		// Now we spend a small amount of time batch reading all available logs
+		// on the log/error channel, the reason we do this is to make sure we read
+		// everything available for the initial batch of logs (when following) or
+		// we read everything available when not following. It's done like this to
+		// ensure that output is consistent and sorted correctly by timestamp when
+		// multiple log-sources are in play.
+		timeout := time.After(time.Millisecond * 25)
+		for {
+			var terminate bool
+			select {
+			case log := <-c:
+				logs = append(logs, log)
+			case err := <-e:
+				if err != io.EOF {
+					logger.Noticef("cannot decode systemd log: %v", err)
+				}
+
+				logReadersDone++
+				if logReadersDone == len(rr.readers) {
+					// break early as this indicates all log sources
+					// have now been closed, which is relevant when doing
+					// no-follow
+					terminate = true
+				}
+			case <-timeout:
+				terminate = true
+			}
+
+			if terminate {
+				break
+			}
+		}
+
+		if err := writeLogs(logs); err != nil {
+			writeError(err)
+		}
+
+		if logReadersDone == len(rr.readers) {
+			break
+		}
 	}
+
 	if err := writer.Flush(); err != nil {
 		logger.Noticef("cannot stream response; problem writing: %v", err)
 	}
-	rr.Close()
```

Comment:
I don't see a place where we close each readers either

Reply:
line 290 is where close is deferred for each reader